src/NHibernate/Collection/Generic/SetHelpers/SetSnapShot.cs

```diff
@@ -153,12 +157,26 @@ protected SetSnapShot(SerializationInfo info, StreamingContext context) : base(i
 
 		void ICollection.CopyTo(Array array, int index)
 		{
-			if (!(array is T[] typedArray))
+			if (array is T[] typedArray)
 			{
-				throw new ArgumentException($"Array must be of type {typeof(T[])}.", nameof(array));
+				CopyTo(typedArray, index);
+				return;
 			}
 
-			CopyTo(typedArray, index);
+			if (array == null)
+				throw new ArgumentNullException(nameof(array));
+
+			if (index < 0)
+				throw new ArgumentOutOfRangeException(nameof(index), index, "Array index cannot be negative");
+
+			if (index > array.Length || Count > array.Length - index)
+				throw new ArgumentException("Provided array is too small", nameof(array));
+
+			foreach (var value in this)
+			{
+				array.SetValue(value, index);
+				index++;
```

Comment:
```suggestion
				array.SetValue(value, index++);
```

Reply:
I avoid combining in one line of code two unrelated operations. That post-increment feature which allows to use a value then increment it altogether looks as an anti-pattern to me.